src/sknnr/_base.py

```diff
@@ -1,19 +1,47 @@
 import numpy as np
-from sklearn.neighbors import KNeighborsRegressor
+from sklearn.base import TransformerMixin
 from sklearn.utils.validation import check_is_fitted
 
 
-class IDNeighborsRegressor(KNeighborsRegressor):
+class KNeighborsDFIndexCrosswalkMixin:
     """
-    Placeholder class for implementing plot ID access.
+    Mixin for KNeighbors estimators that crosswalk array indices to dataframe indexes.
     """
 
+    def fit(self, X, y):
+        """Store dataframe indexes if X is a dataframe."""
+        if hasattr(X, "index"):
+            self.dataframe_index_in_ = np.asarray(X.index)
+
+        return super().fit(X, y)
+
+    def kneighbors(
+        self,
+        X=None,
+        n_neighbors=None,
+        return_distance=True,
+        return_dataframe_index=False,
+    ):
+        """Override kneighbors to optionally return dataframe indexes."""
+        neigh_dist, neigh_ind = super().kneighbors(
+            X=X, n_neighbors=n_neighbors, return_distance=True
+        )
 
-class TransformedKNeighborsMixin(KNeighborsRegressor):
+        if return_dataframe_index:
+            msg = "Dataframe indexes can only be returned when fitted with a dataframe."
+            check_is_fitted(self, "dataframe_index_in_", msg=msg)
+            neigh_ind = self.dataframe_index_in_[neigh_ind]
+
+        return (neigh_dist, neigh_ind) if return_distance else neigh_ind
+
+
+class TransformedKNeighborsMixin:
     """
     Mixin for KNeighbors regressors that apply transformations to the feature data.
     """
 
+    transform_: TransformerMixin
```

Comment:
Without this, `mypy` complained when we used `transform_` because it thought the attribute didn't exist. That wasn't a problem before because we inherited from an untyped `sklearn` class, which `mypy` just assumed had that attribute. 

Maybe there's a better way to fix this? We could just use `type: ignore`, but we'd have to do it in a few different places.

Reply:
The only thing that I don't like about it is that the class doesn't actually have that attribute until it's fit, but I agree that it's nice to explicitly document the attribute and it's type, and I don't have any better ideas to solve it. If you're okay with it, I'll leave it as is.